### PR TITLE
Event PubSub topics + linear filtering.

### PR DESCRIPTION
**Pull request by calavera**
Although these two changes could go separatedly, I rather put them together because they are very related. Our filtering system is very inneficient at matching results. This can slow down the events API with complicated filters. Moreover, that API performs operations in linear complexity at best, and exponencial at worst, when we can do constant time at best and linear at worst.

I separated both commits and wrote explanation of the changes in each of them, but I'm copying both messages here for easy reading:

```
Add PubSub topics.

A TopicFunc is an interface to let the pubisher decide whether it needs to
send a message to a subscriber or not. It returns true if the publisher
must send the message and false otherwise.

Users of the pubsub package can create a subscriber with a topic function
by calling `pubsub.SubscribeTopic`.

Message delivery has also been modified to use concurrent channels per
subscriber. That way, topic verification and message delivery is not o(N+M)
anymore, based on the number of subscribers and topic verification
complexity.

Using pubsub topics, the API stops controlling the message delivery,
delegating that function to a topic generated with the filtering provided
by the user. The publisher sends every message to the subscriber if there
is no filter, but the api doesn't have to select messages to return
anymore.
```

```
Make filtering a linear operation.

Improves the current filtering implementation complixity. Currently, the
best case is O(N) and worst case O(N^2) for key-value filtering. In the new
implementation, the best case is O(1) and worst case O(N), again for
key-value filtering.
```

### CODE REVIEW

**Comment by thaJeztah** [comment]
Nice! Thanks @calavera

**Comment by LK4D4** [comment]
Didn't know that we have crappy `cancel` function :/

**Comment by calavera** [comment]
@LK4D4 we can probably change it. That `cancel` function only calls `daemon.events.Evict` with the listener as an argument. I think it would be much more clean if we had an `Unsubscribe` function in the daemon that took the listener.

That way, the events interface in the daemon could be something like:

```go
type eventsBackend interface {
   Subscribe(since, until int64, eventFilter filterArgs) ([]interface{}, chan interface{})
   Unsubscribe(listener chan interface{})
}
```

**Comment by calavera** [comment]
@LK4D4 that `cancel` crappy function is gone now :smile:  can you take a look?

File: pkg/pubsub/publisher.go [hunk 1]
```diff
@@ -82,3 +81,23 @@ func (p *Publisher) Close() {
...
+		return
+	}
+
+	// send under a select as to not block if the receiver is unavailable
+	if p.timeout > 0 {
+		select {
+		case sub <- v:
+		case <-time.After(p.timeout):
+		}
+	} else {
```
Comment (LK4D4) [inline pkg/pubsub/publisher.go]: You can eliminate else now :)

Comment (calavera) [inline pkg/pubsub/publisher.go]: done, thanks :)

**Comment by LK4D4** [comment]
LGTM

**Comment by tonistiigi** [comment]
LGTM

**Comment by tiborvass** [comment]
LGTM

**Comment by twhiteman** [comment]
Gah - please be careful - your changing/breaking docker remote API when changing raw API types (which affects docker compose and other tools), and there's been no update to the docs (API or release notes) for this change:
https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.23.md#list-containers

Old remote API format: `filters={"exited":["0"]}`
New remote API format: `filters={"exited":{"0":true}}`

**Comment by calavera** [comment]
@twhiteman filters are backwards compatible and the api knows how to deserialize both formats. See https://github.com/docker/engine-api/blob/master/types/filters/parse.go#L93-L115.

Please, feel free to open an issue if you bumped into problems.

**Pull request merged**